expression/builtin_regexp.go

```diff
@@ -1096,43 +1116,191 @@ func (re *builtinRegexpReplaceFuncSig) Clone() builtinFunc {
+	if len(bexpr) == 0 {
+		replacedBStr, err := re.processEmptyStr(reg, instructions, pos, occurrence)
+		if err != nil {
+			return "", false, err
+		}
+		return fmt.Sprintf("0x%s", strings.ToUpper(hex.EncodeToString(replacedBStr))), false, nil
 	}
 
-	replacedBStr := reg.ReplaceAllFunc(trimmedBexpr, repFunc)
+	if occurrence == 0 {
+		replacedStr, err := re.replaceAllMatchedBinStr(reg, bexpr, trimmedBexpr, instructions, pos)
+		if err != nil {
+			return "", false, err
+		}
+		return fmt.Sprintf("0x%s", strings.ToUpper(hex.EncodeToString(replacedStr))), false, nil
+	}
 
-	return fmt.Sprintf("0x%s", strings.ToUpper(hex.EncodeToString(append(bexpr[:pos-1], replacedBStr...)))), false, nil
+	replacedStr, err := re.replaceOneMatchedBinStr(reg, bexpr, trimmedBexpr, instructions, pos, occurrence)
+	if err != nil {
+		return "", false, err
+	}
+	return fmt.Sprintf("0x%s", strings.ToUpper(hex.EncodeToString(replacedStr))), false, nil
 }
 
-func (re *builtinRegexpReplaceFuncSig) getReplacedStr(reg *regexp.Regexp, expr string, trimmedExpr string, repl string, trimmedLen int64, occurrence int64) (string, bool, error) {
-	count := occurrence
-	repFunc := func(matchedStr string) string {
-		if occurrence == 0 {
-			return repl
+func (re *builtinRegexpReplaceFuncSig) getReplacedStr(reg *regexp.Regexp, expr string, trimmedExpr string, instructions []Instruction, trimmedLen int64, occurrence int64) (string, bool, error) {
+	if len(expr) == 0 {
+		replacedStr, err := re.processEmptyStr(reg, instructions, trimmedLen, occurrence)
+		if err != nil {
+			return "", false, nil
 		}
+		return string(replacedStr), false, nil
+	}
 
-		count--
-		if count == 0 {
-			return repl
+	if occurrence == 0 {
+		return re.replaceAllMatchedStr(reg, expr, trimmedExpr, instructions, trimmedLen)
+	}
+
+	return re.replaceOneMatchedStr(reg, expr, trimmedExpr, instructions, trimmedLen, occurrence)
+}
+
+func getInstructions(repl []byte) ([]Instruction, error) {
+	instructions := make([]Instruction, 0)
+	var literals []byte
+
+	replLen := len(repl)
+	for i := 0; i < replLen; i += 1 {
+		if repl[i] == '$' && i+1 < replLen {
+			if stringutil.IsNumericASCII(repl[i+1]) { // Substitution
+				if len(literals) != 0 {
+					instructions = append(instructions, Instruction{SubstitutionNum: -1, Literal: literals})
+					literals = []byte{}
+				}
+				instructions = append(instructions, Instruction{SubstitutionNum: int(repl[i+1] - '0')})
+			} else {
+				literals = append(literals, repl[i+1]) // Escaping
+			}
+			i += 1
+		} else {
+			literals = append(literals, repl[i]) // Plain character
 		}
+	}
 
-		return matchedStr
+	if len(literals) != 0 {
+		instructions = append(instructions, Instruction{SubstitutionNum: -1, Literal: literals})
 	}
 
-	replacedStr := reg.ReplaceAllStringFunc(trimmedExpr, repFunc)
-	return expr[:trimmedLen] + replacedStr, false, nil
+	for _, instruction := range instructions {
+		if instruction.SubstitutionNum >= maxSubExprNum {
```

Comment:
Since when init `SubstitutionNum` as `SubstitutionNum = int(repl[i+1] - '0')` only happens if `stringutil.IsNumericASCII(repl[i+1])` is true and `maxSubExprNum` is hard coded as `10`, it looks to me that `instruction.SubstitutionNum >= maxSubExprNum` is always false

Reply:
> Since when init `SubstitutionNum` as `SubstitutionNum = int(repl[i+1] - '0')` only happens if `stringutil.IsNumericASCII(repl[i+1])` is true and `maxSubExprNum` is hard coded as `10`, it looks to me that `instruction.SubstitutionNum >= maxSubExprNum` is always false

I will remove `maxSubExprNum`